Add blog post on MongoDB sorting for scalar values (#3200)
FerretDB/FerretDB: website/blog/2023-08-14-mongodb-sorting-scalar.md

| 5                                           | Array                                        |
| 6                                           | BinData                                      |
| 7                                           | ObjectId                                     |
| 8                                           | Boolean                                      |
| 9                                           | Date                                         |
| 10                                          | Timestamp                                    |
| 11                                          | Regular Expression                           |

## Comparison of different BSON types

For comparing different BSON types, each BSON type has predefined order of comparison assigned.
Null BSON type has the lowest order of comparison, so Null is the lowest BSON value.
When you compare Null with String, Null is less than any String.

Comparing different BSON type is merely looking up the order of comparison table for each BSON type and using predefined order.
There are exceptions for Object and Array values, we will be discussing them in another blog post.

> ## Number comparison

Author: Thanks for the feedback, addressed it in the blog :hugs:

---
slug: mongodb-sorting-scalar
title: How MongoDB Sorting Works for Scalar Values
authors: [chi]
description: >
  In this blog post, we explore how MongoDB sorting works for scalar values.
keywords:
  - MongoDB sorting
  - BSON comparison order
image: /img/blog/post-cover-image.jpg
tags: [community, product, tutorial]
---

![MongoDB sorting works for scalar values](/img/blog/mongodb-sorting-scalar.jpg)

In this blog post, we explore how MongoDB sorting works for scalar values.

<!--truncate-->

> Sorting compares BSON values to determine which value is equal, greater or less than the other to order them in the ascending or descending order.

Reviewer: ```suggestion
Sorting compares BSON values to determine which value is equal, greater or less than the other to return them in the ascending or descending order.
```

Maybe "to return them" or "to arrange them", so we don't have the word "order" twice?

For comparing different BSON types, [BSON comparison order](#bson-comparison-order) is used.

## BSON comparison order

If two BSON values share the same BSON type, their values are compared to determine which value is greater or less.

> However, if the BSON types are different, a predefined BSON comparison order is used to determine which BSON type is greater or less.

Reviewer: I'm not sure it's fully clear when we say "which BSON type is greater or less"... We still want to understand which value is greater or less, but as they are of different types, we use comparison order for types first. Probably the current statement is also straightforward having in mind the rest of the context, so feel free to ignore my comment.

| Order of Comparison<br/>(lowest to highest) | BSON Types                               |
| ------------------------------------------- | ---------------------------------------- |
| 1                                           | Null                                     |
| 2                                           | Numbers (Integer, Long, Double, Decimal) |
| 3                                           | String                                   |
| 4                                           | Object                                   |
| 5                                           | Array                                    |
| 6                                           | BinData                                  |
| 7                                           | ObjectId                                 |
| 8                                           | Boolean                                  |
| 9                                           | Date                                     |
| 10                                          | Timestamp                                |
| 11                                          | Regular Expression                       |

> ## Comparison of different BSON types

Reviewer: ```suggestion
## Comparison of values with different BSON types
```

(Same as my comment above)

> For comparing different BSON types, each BSON type has a predefined order of comparison assigned.

Reviewer: ```suggestion
To compare values of different BSON types, we need to have in mind that each BSON type has a predefined order of comparison assigned.
```

Boolean BSON type on the other hand has the order of comparison 8, a BSON type such as ObjectId with a lower order of comparison is less than a Boolean value
and a BSON type such as Timestamp with a higher order of comparison is greater than a Boolean value.

Comparing different BSON types is merely looking up the order of comparison table for each BSON type and comparing the predefined order.
There is an exception for Array values, we will be discussing them in another blog post.

### Number comparison

Although numbers have different BSON types, namely Integer, Long, Double and Decimal, they are considered the equivalent BSON type for the purpose of comparison.
That means comparing numbers consider their values but whether they are Integer, Long, Double or Decimal are not relevant.
For instance, an Integer value 0 is equivalent to Double 0.0 as far as comparison is concerned.

### Null and non-existent field comparison

For the comparison purpose, non-existent field is equivalent to Null.
This means that Null and non-existent field are equal as far as comparison is concerned.

## Examples showcasing Sorting for scalar values

> Suppose there is an `outfits` collection, following query inserts documents.

Reviewer: In principle, we don't need this collection to exist to be able to insert documents. Rather the opposite, we need this collection to only contain the documents we insert :) as otherwise we don't know how the order will be affected by the documents existed in the collection before.

```js
db.outfits.insertMany([
  { _id: 1, name: 'flip flops', size: 'M', color: 'blue' },
  { _id: 2, name: 'sandals', size: 9, color: null },
  { _id: 3, name: 'boots', size: 8, color: 'black' },
  { _id: 4, name: 'sneakers', size: 8.5, color: 'blue' },
  { _id: 5, name: 'slippers' }
])
```

The `outfits` collection has a `size` field, and it contains different BSON types.
The document `flip flops` has String field value, `sandals` and `boots` have Integer field values,
`sneakers` has Double field value and `slippers` is missing the field `size`.

> To sort the collection in ascending order by `size` field, sorting order 1 is used and the following query is run.

Reviewer: We don't sort the collection, we return the queried data in ascending order, but it doesn't modify the collection itself.

```js
db.outfits.find().sort({ size: 1 })
```

```json5
[
  { _id: 5, name: 'slippers' },
  { _id: 3, name: 'boots', size: 8, color: 'black' },
  { _id: 4, name: 'sneakers', size: 8.5, color: 'blue' },
  { _id: 2, name: 'sandals', size: 9, color: null },
  { _id: 1, name: 'flip flops', size: 'M', color: 'blue' }
]

> ```

Reviewer: I like this example a lot! It really demonstrates the difference when we compare values of different types!
